trestle/core/control_interface.py

```diff
@@ -0,0 +1,616 @@
+# Copyright (c) 2022 IBM Corp. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Handle queries and utility operations on controls in memory."""
+from __future__ import annotations
+
+import copy
+import logging
+import pathlib
+import re
+from enum import Enum
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+import trestle.oscal.catalog as cat
+from trestle.common import const
+from trestle.common.common_types import TypeWithProps
+from trestle.common.err import TrestleError
+from trestle.common.list_utils import as_dict, as_list, none_if_empty
+from trestle.common.str_utils import string_from_root
+from trestle.core.trestle_base_model import TrestleBaseModel
+from trestle.oscal import common
+from trestle.oscal import component as comp
+from trestle.oscal import profile as prof
+
+logger = logging.getLogger(__name__)
+
+
+class ParameterRep(Enum):
+    """Enum for ways to represent a parameter."""
+
+    LEAVE_MOUSTACHE = 0
+    VALUE_OR_STRING_NONE = 1
+    LABEL_OR_CHOICES = 2
+    VALUE_OR_LABEL_OR_CHOICES = 3
+    VALUE_OR_EMPTY_STRING = 4
+
+
+class ComponentImpInfo(TrestleBaseModel):
+    """Class to capture component prose and status."""
+
+    prose: str
+    status = common.ImplementationStatus(state=const.STATUS_OTHER)
+
+
+# provide name for this type
+CompDict = Dict[str, Dict[str, ComponentImpInfo]]
+
+
+class ContextPurpose(Enum):
+    """Specify the modality of the control markdown."""
+
+    CATALOG = 0
+    COMPONENT = 1
+    PROFILE = 2
+    SSP = 3
+
+
+class ControlContext(TrestleBaseModel):
```

Comment:
I don't think we should use `TrestleBaseModel` here, `TrestleBaseModel` services as a base for all OSCAL model classes and does some validation on the class initialisation when we load some OSCAL model. ControlContext should not be a subclass of this class.

I think it is rather debatable whether we should use pydantic here. The context initialisation is done only by us, internally, and invoking pydantic with all its validation and background work seems to be an overkill here. We can define all the fields in the __init__ function and treat it as normal class.